File: godci/src/v6/domains/accounts.go
package domains

import "time"

type AccountInfo struct {
	//Struct used to capture account metadata.
	VendorName      string
	AccountID       int
	UserID          int
	StartingBalance float64
	AccountType     string
	ProductCategory string
}

type EntryItem struct {
	//Struct used to capture account entries.
	AccountID       int
	Date            time.Time
	Message         string
	Amount          float64
	TransactionType string
}

type AccountDomain struct {
	//Struct used to represent an account. It is composed of static account
	//metadata and entries that record the transaction history of the
	//account.
	*AccountInfo
	Entries []*EntryItem
}

func (accDomain *AccountDomain) Balance() float64 {
	//Balance calculates the current balance of an account from the entries of the
	// account.
	currentBalance := accDomain.StartingBalance
	for _, entry := range accDomain.Entries {
		switch entry.TransactionType {
		case CREDIT:
			currentBalance += entry.Amount
		case DEBIT:
			currentBalance -= entry.Amount
		}
	}
	return currentBalance
}

func (accDomain *AccountDomain) HasSufficentFunds(amount float64) bool {
	//HasSufficentFunds determines if the account has sufficent funds to
	//cover the amount provided by the calling entity.
	if accDomain.Balance() < amount {
		return false
	}
	return true
}

func (ad *AccountDomain) RecordTransaction(message string, date time.Time, amt float64, transtype string) {
	//RecordTransaction records an entry in an account.
	er := &EntryItem{
		AccountID:       ad.AccountID,
		Date:            date,
		Amount:          amt,
		TransactionType: transtype,
		Message:         message,
	}
	ad.Entries = append(ad.Entries, er)
}

File: jsdci/v8/domains/accounts.js
(function () {
  "use strict";

  var Constants = require('./constants.js');

  var AccountInfo, EntryItem, AccountDomain, createAccount;

  AccountInfo = {
    init: function init(accountID, lastName, firstName, startingBalance, accountType) {
      this.accountID = accountID;
      this.lastName = lastName;
      this.firstName = firstName;
      this.startingBalance = startingBalance;
      this.accountType = accountType;
    }
  };

  EntryItem = {
    init: function init(accountID, date, message, amount, transactionType) {
      this.accountID = accountID;
      this.date = date;
      this.message = message;
      this.amount = amount;
      this.transactionType = transactionType;
    }
  };

  AccountDomain = {
    init: function init(accountInfo, entryItems) {
      this.accountInfo = accountInfo;
      this.entries = entryItems;
    },
    balance: function balance() {
      var entry, currentBalance, indexer, len;
      currentBalance = this.accountInfo.startingBalance * 100;
      for (indexer = 0, len = this.entries.length; indexer < len; indexer += 1) {
        entry = this.entries[indexer];
        switch (entry.transactionType) {
          case Constants.Credit:
            currentBalance += entry.amount * 100;
            break;
          case Constants.Debit:
            currentBalance -= entry.amount * 100;
            break;
        }
      }
      return currentBalance/100;
    }
  };

  createAccount = function createAccount(id, lastName, firstName, startingBalance, accountType) {
    var accountInfo, entries, accountDomain;

    accountInfo = Object.create(AccountInfo);
    accountInfo.init(id, lastName, firstName, startingBalance, accountType);
    entries =[];
    accountDomain = Object.create(AccountDomain);
    accountDomain.init(accountInfo, entries);

    return accountDomain;
  };

  module.exports = {AccountInfo: AccountInfo, EntryItem: EntryItem,
    AccountDomain: AccountDomain, createAccount: createAccount
  };
})();
